test/integration/acceptance/fake_peer_example_test.cpp

```diff
@@ -318,80 +330,98 @@ TEST_F(FakePeerExampleFixture,
+ * @when fake peer provides a proposal with valid tx
+ * @then the real peer must commit the transaction from that proposal
  */
 TEST_F(FakePeerExampleFixture,
        OnDemandOrderingProposalAfterValidCommandReceived) {
 
-  /* A custom behaviour that requests a proposal for the round it got vote for,
-   * and if gets one, checks that the proposal contains the given tx hash.
-   */
-  struct CustomBehaviour : public fake_peer::HonestBehaviour {
-    CustomBehaviour(const interface::types::HashType &tx_hash,
-                    std::atomic_flag &got_proposal_from_main_peer)
-        : tx_hash_(tx_hash),
-          got_proposal_from_main_peer_(got_proposal_from_main_peer) {}
-
-    void processYacMessage(fake_peer::YacMessagePtr message) override {
-      const auto proposal_from_main_peer = getFakePeer().sendProposalRequest(
-          message->front().hash.vote_round, kProposalWaitingTime);
-      if (proposal_from_main_peer
-          and std::any_of(proposal_from_main_peer->transactions().begin(),
-                          proposal_from_main_peer->transactions().end(),
-                          [this](const auto &tx) {
-                            return tx.reducedHash() == tx_hash_;
-                          })) {
-        got_proposal_from_main_peer_.test_and_set(std::memory_order_relaxed);
-      }
-      HonestBehaviour::processYacMessage(message);
-    }
-
-    const interface::types::HashType &tx_hash_;
-    std::atomic_flag &got_proposal_from_main_peer_;
-  };
-
   // Create the tx:
   const auto tx = complete(
       baseTx(kAdminId).transferAsset(kAdminId, kUserId, kAssetId, "tx1", "1.0"),
       kAdminKeypair);
-  const auto hash = tx.reducedHash();
 
-  std::atomic_flag got_proposal_from_main_peer = ATOMIC_FLAG_INIT;
+  auto state_ready = std::make_shared<std::atomic_bool>(false);
 
-  auto &itf = prepareState(1);
-  fake_peers_.front()->setBehaviour(
-      std::make_shared<CustomBehaviour>(hash, got_proposal_from_main_peer));
+  auto proposal_storage = std::make_shared<fake_peer::ProposalStorage>();
+  proposal_storage->setDefaultProvider(
+      [tx, state_ready](
+          const auto &round) -> fake_peer::OrderingProposalRequestResult {
+        if (!state_ready->load(std::memory_order_relaxed)) {
+          return boost::none;
+        }
+        shared_model::proto::ProtoProposalFactory<
+            validation::AlwaysValidValidator>
+            valid_factory;
+        std::shared_ptr<const shared_model::proto::Proposal> proposal;
+        valid_factory
+            .createProposal(round.block_round,
+                            iroha::time::now(),
+                            std::vector<shared_model::proto::Transaction>({tx}))
+            .match(
+                [&proposal](iroha::expected::Value<std::unique_ptr<
+                                shared_model::interface::Proposal>> &v) {
+                  auto p1 =
+                      std::static_pointer_cast<shared_model::proto::Proposal>(
```

Comment:
D: may you a bit elaborate what is happening here?

Reply:
That was a debugging code I forgot to remove, sorry.